server/src/main/java/org/apache/druid/curator/CuratorModule.java

```diff
@@ -63,15 +63,20 @@
   @Override
   public void configure(Binder binder)
   {
+    JsonConfigProvider.bind(binder, CURATOR_CONFIG_PREFIX, ZkEnablementConfig.class);
     JsonConfigProvider.bind(binder, CURATOR_CONFIG_PREFIX, CuratorConfig.class);
     JsonConfigProvider.bind(binder, EXHIBITOR_CONFIG_PREFIX, ExhibitorConfig.class);
   }
 
   @Provides
   @LazySingleton
   @SuppressForbidden(reason = "System#err")
-  public CuratorFramework makeCurator(CuratorConfig config, EnsembleProvider ensembleProvider, Lifecycle lifecycle)
+  public CuratorFramework makeCurator(ZkEnablementConfig zkEnablementConfig, CuratorConfig config, EnsembleProvider ensembleProvider, Lifecycle lifecycle)
```

Comment:
I haven't looked too closely yet, but I wonder if this would be better if this was marked as `@Nullable` and returned null instead of throwing the runtime exception, and shift the burden of validating that curator is available to the settings that do require it, such as inventory, segment loading, and task management? The other stuff might be able to be simplified a bit and not have to care about having the setting, and could probably avoid having some of the signature changes to use providers.

Reply:
Main reason for failing loudly here is so that if I forgot to disable zk in some code  path, then this method would immediately  fail on node start with  guice injection errors leading to  quick discovery of exactly what is missed.
this helped me catch quite a few places that I missed.